src/hogwild.rs

```diff
@@ -0,0 +1,113 @@
+use std::error::Error;
+use std::sync::{Arc, mpsc, Mutex};
+use std::sync::mpsc::{Receiver, Sender};
+use std::thread;
+use std::thread::JoinHandle;
+
+use crate::feature_buffer::{FeatureBuffer, FeatureBufferTranslator};
+use crate::model_instance::ModelInstance;
+use crate::multithread_helpers::BoxedRegressorTrait;
+use crate::port_buffer::PortBuffer;
+use crate::regressor::Regressor;
+
+pub struct HogwildTrainer {
+    workers: Vec<JoinHandle<()>>,
+    sender: Sender<Vec<u32>>,
+}
+
+pub struct HogwildWorker {
+    regressor: BoxedRegressorTrait,
+    feature_buffer_translator: FeatureBufferTranslator,
+    port_buffer: PortBuffer
+}
+
+impl HogwildTrainer {
+    pub fn new(sharable_regressor: BoxedRegressorTrait, model_instance: &ModelInstance, numWorkers: u32) -> HogwildTrainer {
+        let (sender, receiver): (Sender<Vec<u32>>, Receiver<Vec<u32>>) = mpsc::channel();
+        let mut trainer = HogwildTrainer {
+            workers: Vec::new(),
+            sender,
+        };
+        let receiver: Arc<Mutex<Receiver<Vec<u32>>>> = Arc::new(Mutex::new(receiver));
+        let feature_buffer_translator = FeatureBufferTranslator::new(model_instance);
+        let port_buffer = sharable_regressor.new_portbuffer();
+        for i in 0..numWorkers {
+            let worker = HogwildWorker::new(
+                sharable_regressor.clone(),
```

Comment:
`shareable_regressor` sounds as if it's actually shared, but in fact it's cloned?

Reply:
yes - it is cloned but uses a smart pointer that actually, under the hood, shares part of the regressor (the weights) between all the clones. So it is in fact (unsafely) sharable